solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceBaseInputOperator.java

```diff
@@ -0,0 +1,319 @@
+import com.datatorrent.netlet.util.DTThrowable;
+
+@SuppressWarnings("unused")
+public abstract class AbstractSolaceBaseInputOperator<T> extends BaseOperator implements
+    InputOperator, Operator.ActivationListener<Context.OperatorContext>, CheckpointListener
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractSolaceBaseInputOperator.class);
+
+  @NotNull
+  protected JCSMPProperties properties = new JCSMPProperties();
+  protected String connectRetries;
+  protected String reconnectRetries;
+  protected String unackedMessageLimit;
+
+
+  //protected IdempotentStorageManager idempotentStorageManager = new IdempotentStorageManager.FSIdempotentStorageManager();
+  protected FSOpsIdempotentStorageManager idempotentStorageManager = new FSOpsIdempotentStorageManager();
+
+  protected transient JCSMPFactory factory;
+  protected transient JCSMPSession session;
+
+  protected transient Consumer consumer;
+  protected transient Consumer reliableConsumer;
+
+  protected transient int operatorId;
+  protected long windowId;
+  protected transient long lastCompletedWId;
+
+  protected transient int emitCount;
+
+  protected transient volatile boolean DRFailover = false;
+  protected transient volatile boolean TCPDisconnected = false;
+
+  protected transient BlockingQueue<BytesXMLMessage> unackedMessages; // hosts the Solace messages that need to be acked when the streaming window is OK to remove
+  protected LinkedList<Long> inFlightMessageId = new LinkedList<Long>(); //keeps track of all in flight IDs since they are not necessarily sequential
+  protected transient ArrayBlockingQueue<BytesXMLMessage> arrivedTopicMessagesToProcess;
+
+  //protected transient com.solace.dt.operator.DTSolaceOperatorInputOutput.ArrayBlockingQueue<BytesXMLMessage> arrivedMessagesToProcess;
+
+  private transient ReconnectCallbackHandler rcHandler = new ReconnectCallbackHandler();
+
+  private transient CallbackMessageHandler cbHandler = new CallbackMessageHandler();
+
+  int spinMillis;
+
+  protected transient int reconnectRetryMillis = 0;
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+
+    operatorId = context.getId();
+    logger.info("OperatorID from Base class: {}", operatorId);
+    spinMillis = context.getValue(com.datatorrent.api.Context.OperatorContext.SPIN_MILLIS);
+    factory = JCSMPFactory.onlyInstance();
+
+    //Required for HA and DR to try forever if set to "-1"
+    JCSMPChannelProperties channelProperties = (JCSMPChannelProperties)this.properties.getProperty(JCSMPProperties.CLIENT_CHANNEL_PROPERTIES);
+    channelProperties.setConnectRetries(Integer.parseInt(this.connectRetries));
+    channelProperties.setReconnectRetries(Integer.parseInt(this.reconnectRetries));
+
+    reconnectRetryMillis = channelProperties.getReconnectRetryWaitInMillis();
+
+
+    try {
+      session = factory.createSession(this.properties, null, new PrintingSessionEventHandler());
```

Comment:
Is it more appropriate to move this to activate() where you createSession just before connect()?

Reply:
This is an offline call so I think it is ok here. It should also be ok to leave the second argument as null because internally it makes that assignment when it is null.